lib/ecto/repo/model.ex

```diff
@@ -176,6 +176,44 @@ defmodule Ecto.Repo.Model do
   end
 
   @doc """
+  Implementation for `Ecto.Repo.insert_or_update/2`.
+  """
+  def insert_or_update(repo, adapter, changeset = %Ecto.Changeset{}, opts) do
+    case get_state(changeset) do
+      :built  -> insert repo, adapter, changeset, opts
+      :loaded -> update repo, adapter, changeset, opts
+      state   -> raise ArgumentError, "the changeset has an invalid state " <>
+                                      "for Repo.insert_or_update/2: #{state}"
+    end
+  end
+  def insert_or_update(repo, _adapter, _model, _opts) do
+    raise ArgumentError, "giving a model to "
+                         "#{inspect repo}.insert_or_update/2 is not " <>
+                         "supported. Please use an `%Ecto.Changeset{}`\n" <>
+                         Exception.format_stacktrace
+  end
+
+  @doc """
+  Implementation for `Ecto.Repo.insert_or_update!/2`.
+  """
+  def insert_or_update!(repo, adapter, changeset = %Ecto.Changeset{}, opts) do
+    case get_state(changeset) do
+      :built  -> insert! repo, adapter, changeset, opts
+      :loaded -> update! repo, adapter, changeset, opts
+      state   -> raise ArgumentError, "the changeset has an invalid state " <>
+                                      "for Repo.insert_or_update/2: #{state}"
+    end
+  end
+  def insert_or_update!(repo, _adapter, _model, _opts) do
+    raise ArgumentError, "giving a model to "
+                         "#{inspect repo}.insert_or_update/2 is not " <>
+                         "supported. Please use an `%Ecto.Changeset{}`\n" <>
+                         Exception.format_stacktrace
+  end
+
+  defp get_state(%Changeset{model: model}), do: model.__meta__.state
```

Comment:
You can make the `get_state` logic check for the changeset, state and raise. This way, we can share it across the two functions. OR we make `insert_or_update!` call `insert_or_update`.